make/photon/prepare/migration/utils.py

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import sys
+import os
+import yaml
+import click
+import importlib
+from string import Template
+
+def read_conf(path):
+    with open(path) as f:
+        try:
+            d = yaml.safe_load(f)
+        except Exception as e:
+            click.echo("parse config file err, make sure your harbor config version is above 1.8.0", e)
+            exit(-1)
+    return d
+
+def to_module_path(ver):
+    return "migration.versions.{}".format(ver.replace(".","_"))
+
+def search(input_ver: str, target_ver: str) -> list:
+    """
+    Search accept a input version and the target version.
+    Returns the module of migrations in the upgrade path
+    """
+    def helper():
+        nonlocal basedir
```

Comment:
what's the benefit here to use `nonlocal` rather than pass these as parms?

the code outside `helper` does not rely on the change `helper` makes to these variables.

Reply:
This is a closure, the inner function can access the variables outside, in python3 if you wanna access outside variables you must declare it as `nonlocal`.